### PR TITLE
feat: thor btcquote

### PR DESCRIPTION
**Pull request by elmutt**
Thorchain getQuote fully working.

Can be tested locally linked to this web PR:
https://github.com/shapeshift/web/pull/2153

To Test:

- set REACT_APP_MIDGARD_URL in web .env to a known good url midgard (https://midgard.thorswap.net/v2 is pretty good)
- REACT_APP_FEATURE_THOR=true

Run web linked locally and see valid quotes & prices appearing for thorchain pairs (BTC/FOX for example)

### CODE REVIEW

File: packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      .decimalPlaces(0)
+      .toString()
+
+    const memo = makeSwapMemo({
+      buyAssetId: buyAsset.assetId,
+      destinationAddress,
+      limit
+    })
+
+    const adapter = await deps.adapterManager.get('bip122:000000000019d6689c085ae165831e93')
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:77]: use `btcChainId` const

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 2]
```diff
@@ -17,63 +20,24 @@ export const getThorTradeQuote = async ({
   deps: ThorchainSwapperDeps
   input: GetTradeQuoteInput
 }): Promise<TradeQuote<ChainId>> => {
-  const { sellAsset, buyAsset, sellAmount, sellAssetAccountNumber, wallet } = input
+  const { sellAsset, buyAsset, sellAmount, sellAssetAccountNumber, wallet, chainId } = input
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:23]: curious looking at this implementation in regards to `sellAssetAccountNumber`, I am wondering why this is something we need to pass in from the client when it should be able to be determined by the via address/xpub using `getAccount()` in chain adapters. I feel like this would end up needing to happen in the web client in this setup when it would be nice to reduce the need for extra calls from the client if we have the necessary data to derive in lib ideally. Thoughts?

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 3]
```diff
@@ -17,63 +20,24 @@ export const getThorTradeQuote = async ({
...
-      assetReference: sellAssetErc20Address,
-      chainReference,
-      chainNamespace
-    } = fromAssetId(sellAsset.assetId)
+    const { assetReference: sellAssetErc20Address, assetNamespace } = fromAssetId(sellAsset.assetId)
 
-    const isErc20Trade = sellAssetErc20Address.startsWith('0x')
-    const sellAssetChainId = toChainId({ chainReference, chainNamespace })
-    const adapter = deps.adapterManager.get(sellAssetChainId)
+    const isErc20Trade = assetNamespace === 'erc20'
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:33]: Similar story with passing `isErc20Trade`, we have the sellAsset being passed to `getEthThorTxInfo()` and can easily be derived within there. There are just a lot of input args for some of these functions and it would be nice to reduce that input payload a bit if possible. These are more overarching idea than directly related to your PR as it stands, so they don't need to be addressed here, but something to think about

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts (hunk 2 above)
Comment (kaladinlight) [reply packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:23]: Or the `bip44Params` even better

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 4]
```diff
@@ -87,23 +51,98 @@ export const getThorTradeQuote = async ({
...
     const buyAmount = normalizeAmount(bnOrZero(sellAmount).times(rate))
-    const sellAssetTradeFee = bnOrZero(feeData.tradeFee).times(bnOrZero(feeAssetRatio))
 
+    const tradeFee = await estimateTradeFee(deps, buyAsset.assetId)
+    const sellAssetTradeFee = bnOrZero(tradeFee).times(bnOrZero(feeAssetRatio))
     // padding minimum by 1.5 the trade fee to avoid thorchain "not enough to cover fee" errors.
     const minimum = fromBaseUnit(sellAssetTradeFee.times(1.5).toString(), sellAsset.precision)
 
-    return {
+    const commonQuoteFields = {
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:60]: can we type this please

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 5]
```diff
@@ -87,23 +51,98 @@ export const getThorTradeQuote = async ({
...
-      allowanceContract: router,
       buyAsset,
       sellAsset,
-      sellAssetAccountNumber
+      sellAssetAccountNumber,
+      minimum
+    }
+
+    switch (chainId) {
+      case 'eip155:1':
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:73]: constants

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 6]
```diff
@@ -87,23 +51,98 @@ export const getThorTradeQuote = async ({
...
+          })
+
+          return {
+            ...commonQuoteFields,
+            allowanceContract: router,
+            feeData
+          }
+        })()
+
+      case 'bip122:000000000019d6689c085ae165831e93':
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:106]: constants

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 7]
```diff
@@ -87,23 +51,98 @@ export const getThorTradeQuote = async ({
...
       buyAsset,
       sellAsset,
-      sellAssetAccountNumber
+      sellAssetAccountNumber,
+      minimum
+    }
+
+    switch (chainId) {
+      case 'eip155:1':
+        return (async (): Promise<TradeQuote<'eip155:1'>> => {
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:74]: constants

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 8]
```diff
@@ -87,23 +51,98 @@ export const getThorTradeQuote = async ({
...
+
+          return {
+            ...commonQuoteFields,
+            allowanceContract: router,
+            feeData
+          }
+        })()
+
+      case 'bip122:000000000019d6689c085ae165831e93':
+        return (async (): Promise<TradeQuote<'bip122:000000000019d6689c085ae165831e93'>> => {
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:107]: constants

File: packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      buyAsset.precision
+    )
+
+    if (
+      !bnOrZero(expectedBuyAmount).gte(0) ||
+      !(bnOrZero(slippageTolerance).gte(0) && bnOrZero(slippageTolerance).lte(1))
+    )
+      throw new SwapError('[makeTradeTx]: bad expected buy amount or bad slippage tolerance', {
+        code: SwapErrorTypes.BUILD_TRADE_FAILED
+      })
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:64]: think it would be worth breaking out these two checks and return an error explicitly stating the fail case?

Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:64]: we can attach some data to the error

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 10]
```diff
@@ -87,23 +51,98 @@ export const getThorTradeQuote = async ({
...
+      sellAssetAccountNumber,
+      minimum
+    }
+
+    switch (chainId) {
+      case 'eip155:1':
+        return (async (): Promise<TradeQuote<'eip155:1'>> => {
+          const bip44Params = adapter.buildBIP44Params({
+            accountNumber: Number(sellAssetAccountNumber)
+          })
```
[the remark this answers is not in the document]
Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:77]: Ya for sure that might be better

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts (hunk 3 above)
Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:33]: for sure we can probably clan this up a bit with regards to passing some things (bip44 instead of account number etc)

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts (hunk 2 above)
Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:23]: for sure can probably be done that way

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 11]
```diff
@@ -87,23 +63,97 @@ export const getThorTradeQuote = async ({
...
     const buyAmount = normalizeAmount(bnOrZero(sellAmount).times(rate))
-    const sellAssetTradeFee = bnOrZero(feeData.tradeFee).times(bnOrZero(feeAssetRatio))
 
+    const tradeFee = await estimateTradeFee(deps, buyAsset.assetId)
+    const sellAssetTradeFee = bnOrZero(tradeFee).times(bnOrZero(feeAssetRatio))
     // padding minimum by 1.5 the trade fee to avoid thorchain "not enough to cover fee" errors.
     const minimum = fromBaseUnit(sellAssetTradeFee.times(1.5).toString(), sellAsset.precision)
 
-    return {
+    const commonQuoteFields: CommonQuoteFields = {
```
Comment (kaladinlight) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:72]: can just use an `Omit` over the dupe `CommonQuoteFields` type
```suggestion
    const commonQuoteFields: Omit<TradeQuote<ChainId>, 'allowanceContract' | 'feeData'> = {```

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts (hunk 5 above)
Comment (kaladinlight) [reply packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:85]: should be able to use `KnownChainIds` after rebasing with main

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts (hunk 6 above)
Comment (kaladinlight) [reply packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:117]: https://github.com/shapeshift/lib/pull/840#discussion_r916286069

**Comment by elmutt** [comment]
> @elmutt Description/Testing steps pls 🙏🏽 to be able to review this one

FIrst I need to fix how we are initializing swappers and make it async. Then I can do final testing on this PR and wlll update the testing instructions

**Comment by 0xApotheosis** [comment]
Converting to draft until ready to review.

**Review by DaoDev44 — DISMISSED** [review]

**Comment by elmutt** [comment]
Found some issues reviewing with kevin today. Converting back into draft until this & web pr are updated

File: packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    const { data: inboundAddresses } = await thorService.get<InboundResponse[]>(
+      `${deps.midgardUrl}/thorchain/inbound_addresses`
+    )
+
+    const btcInboundAddresses = inboundAddresses.find((inbound) => inbound.chain === 'BTC')
+
+    const vault = btcInboundAddresses?.address
+
+    if (!vault)
+      throw new SwapError(`[getPriceRatio]: vault not found for BTC`, {
```
Comment (DaoDev44) [inline packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:44]:
```suggestion
      throw new SwapError(`[getThorTxInfo]: vault not found for BTC`, {
```

File: packages/swapper/src/api.ts [hunk 13]
```diff
@@ -60,6 +60,7 @@ type CommonTradeInput = {
   sendMax: boolean
   sellAssetAccountNumber: number
   wallet?: HDWallet // TODO remove this in a followup PR
```
Comment (0xdef1cafe) [inline packages/swapper/src/api.ts:62]: why is this optional? is this comment still valid?

File: packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+import { Asset, BIP44Params, UtxoAccountType } from '@shapeshiftoss/types'
+
+import { SwapError, SwapErrorTypes } from '../../../../../api'
+import { bn, bnOrZero, fromBaseUnit, toBaseUnit } from '../../../../utils/bignumber'
+import { InboundResponse, ThorchainSwapperDeps } from '../../../types'
+import { getPriceRatio } from '../../getPriceRatio/getPriceRatio'
+import { makeSwapMemo } from '../../makeSwapMemo/makeSwapMemo'
+import { thorService } from '../../thorService'
+
+export const getThorTxInfo = async ({
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:13]: nicer to type `type GetBtcThorTxInfo = (args: GetBtcThorTxInfoArgs) => GetBtcThorTxInfoReturn`

File: packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    })
+
+    const expectedBuyAmount = toBaseUnit(
+      fromBaseUnit(bnOrZero(sellAmount).dividedBy(priceRatio), sellAsset.precision),
+      buyAsset.precision
+    )
+
+    if (
+      !bnOrZero(expectedBuyAmount).gte(0) ||
+      !(bnOrZero(slippageTolerance).gte(0) && bnOrZero(slippageTolerance).lte(1))
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:61]: change first line to `bnOrZero(expectedBuyAmount).lt(0)` and name the second line something like `isValidSlippageRange`

File: packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts (hunk 14 above)
Comment (0xdef1cafe) [reply packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:13]: rename this to `getBTCThorTxInfo` and the existing other one to `getETHThorTxInfo`

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 16]
```diff
@@ -87,23 +59,88 @@ export const getThorTradeQuote = async ({
...
     const buyAmount = normalizeAmount(bnOrZero(sellAmount).times(rate))
-    const sellAssetTradeFee = bnOrZero(feeData.tradeFee).times(bnOrZero(feeAssetRatio))
 
+    const tradeFee = await estimateTradeFee(deps, buyAsset.assetId)
+    const sellAssetTradeFee = bnOrZero(tradeFee).times(bnOrZero(feeAssetRatio))
     // padding minimum by 1.5 the trade fee to avoid thorchain "not enough to cover fee" errors.
     const minimum = fromBaseUnit(sellAssetTradeFee.times(1.5).toString(), sellAsset.precision)
 
-    return {
+    const commonQuoteFields: Omit<TradeQuote<ChainId>, 'allowanceContract' | 'feeData'> = {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:68]: can we abstract this to `type CommonQuoteFields = Omit<TradeQuote<ChainId>, 'allowanceContract' | 'feeData'>` and use it?

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 17]
```diff
@@ -87,23 +59,88 @@ export const getThorTradeQuote = async ({
     const priceRatio = await getPriceRatio(deps, { sellAssetId, buyAssetId })
     const rate = bnOrZero(1).div(priceRatio).toString()
     const buyAmount = normalizeAmount(bnOrZero(sellAmount).times(rate))
-    const sellAssetTradeFee = bnOrZero(feeData.tradeFee).times(bnOrZero(feeAssetRatio))
 
+    const tradeFee = await estimateTradeFee(deps, buyAsset.assetId)
+    const sellAssetTradeFee = bnOrZero(tradeFee).times(bnOrZero(feeAssetRatio))
     // padding minimum by 1.5 the trade fee to avoid thorchain "not enough to cover fee" errors.
     const minimum = fromBaseUnit(sellAssetTradeFee.times(1.5).toString(), sellAsset.precision)
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts:66]: not part of this PR, but 1.5 seems excessive?

File: packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+  accountType
+}: {
+  deps: ThorchainSwapperDeps
+  sellAsset: Asset
+  buyAsset: Asset
+  sellAmount: string
+  slippageTolerance: string
+  destinationAddress: string
+  wallet: HDWallet
+  bip44Params: BIP44Params
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:31]: do we need to pass these in?

https://github.com/shapeshift/lib/blob/main/packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.ts#L44

File: packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+      buyAssetId: buyAsset.assetId,
+      destinationAddress,
+      limit
+    })
+
+    const adapter = deps.adapterManager.get(btcChainId)
+
+    const pubkey = await (adapter as unknown as bitcoin.ChainAdapter).getPublicKey(
+      wallet,
+      bip44Params,
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/bitcoin/utils/getThorTxData.ts:83]: https://github.com/shapeshift/lib/blob/main/packages/chain-adapters/src/bitcoin/BitcoinChainAdapter.ts#L267

this might help? if we're just overriding `accountNumber`?

File: packages/swapper/src/api.ts (hunk 13 above)
Comment (elmutt) [reply packages/swapper/src/api.ts]: fixed

**Review by DaoDev44 — DISMISSED** [review]

**Review by 0xdef1cafe — APPROVED** [review]
